### PR TITLE
Sabrina/annotate skip

### PR DESCRIPTION
**Pull request by sabrina-li**
https://github.com/fullstorydev/fullstory-babel-plugin-annotate-react/issues/33#issuecomment-923267797

Add a way to use comments to skip annotation for certain components. This could be useful when certain component (such as material-ui's `ThemeProvider`) does it's own run time `prop-type` checks and throw warnings on the `data-*` attributes that we add.

### CODE REVIEW

**Comment by JoshMiers-FS** [comment]
The issue I am referring to is `react-nativgation`: https://github.com/react-navigation/react-navigation/blob/2.x/src/createNavigationContainer.js#L24-L32

File: README.md [hunk 1]
```diff
@@ -68,6 +68,26 @@ If you would like the plugin to attempt to annotate the first HTML element creat
...
+```javascript
+  plugins: [
+      [
+        '../..',
+        { 
+          "ignore-components":[
+            // each item must be a string array containing three items: file name, component name, element name 
+            // corresponding to the values for data-source-file, data-component, data-element
+            // use wild card (*) to match anything
+            ["myBoxComponent.jsx","MyBox","Box"],
```
Comment (JoshMiers-FS) [inline README.md:82]: Nice. This is exactly what I was thinking we would need to provide so that we can handle scenarios where only the component name is know and scenarios where there is a collision and more info needs to be added.

File: index.js [hunk 2]
```diff
@@ -149,14 +173,15 @@ function applyAttributes(t, openingElement, componentName, sourceFileName, compo
...
+  if (
     sourceFileName
+    && !ignoredComponentFromOptions
     && (componentName || ignoredElement === false)
     && openingElement.node.attributes.find(node => {
       if (!node.name) return
       return node.name.name === sourceFileAttributeName
     }
-  ) == null){
+  ) == undefined){
```
Comment (JoshMiers-FS) [inline index.js:184]: why this change?

File: index.js [hunk 3]
```diff
@@ -20,7 +24,8 @@ module.exports = function({ types: t }) {
           path,
           path.node.id.name,
           sourceFileNameFromState(state),
-          ...attributeNamesFromState(state)
```
Comment (JoshMiers-FS) [inline index.js:23]: what's the significance of removing the `...`?

Comment (sabrina-li) [reply index.js:23]: tbh not much, just that the parameters for these functions got too long. Also adding the new `ignoreComponentsFromOption` after a spread operator seemed like it could be confusing. I don't feel really strongly about using `...` here, maybe the `ignoreComponentsFromOption` can come before the `sourceFileNameFromState` and the `...attributeNamesFromState(state)` can remain in the last position

File: index.js (hunk 2 above)
Comment (sabrina-li) [reply index.js:184]: Ah good point! I think I had a bug in line 137 where I was strictly checking `!== null` but `find` returns `undefined` when no match.
So I probably overkilled by changing all of the `.find` compares without realizing these aren't type matching. For consistency I can update line 137 to do `!= null` instead

File: .circleci/config.yml [hunk 4]
```diff
@@ -22,6 +22,9 @@ jobs:
       - run:
           name: Audit
           command: npm audit --audit-level=low
+      - run:
+          name: Test
+          command: npm run test
```
Comment (nfriedly) [inline .circleci/config.yml:27]: 🎉

File: index.js [hunk 5]
```diff
@@ -106,7 +115,8 @@ function isReactFragment(openingElement) {
   )
 }
 
-function applyAttributes(t, openingElement, componentName, sourceFileName, componentAttributeName, elementAttributeName, sourceFileAttributeName) {
+function applyAttributes(t, openingElement, componentName, sourceFileName, attributeNames, ignoreComponentsFromOption) {
+  const [ componentAttributeName, elementAttributeName, sourceFileAttributeName ] = attributeNames;
```
Comment (nfriedly) [inline index.js:119]: @JoshMiers-FS if it wasn't clear, the `...` above is replaced by this line here. They both have the effect of extracting each of the values of the array into their own variables. See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax for more info about `...` and https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment for more about this replacement.

File: index.js [hunk 6]
```diff
@@ -116,30 +126,44 @@ function applyAttributes(t, openingElement, componentName, sourceFileName, compo
...
 
+  const elementName = openingElement.node.name.name || 'unknown'
+  const ignoredComponentFromOptions = ignoreComponentsFromOption && ignoreComponentsFromOption.find(component => {
+    let match = true;
+    if (component[0] !== '*' && component[0] !== sourceFileName) match = false;
+    else if (component[1] !== '*' && component[1] !== componentName) match = false;
+    else if (component[2] !== '*' && component[2] !== elementName) match = false;
+    
+    return match
+  }) !== undefined
```
Comment (nfriedly) [inline index.js:137]: This feels a little confusing.. but I'm not sure I have a good suggestion for making it better. I believe the logic is correct, after thinking about it for a bit.

Comment (nfriedly) [reply index.js:137]: Maybe do `!!` instead of comparing to undefined?

File: index.js [hunk 7]
```diff
@@ -116,30 +126,44 @@ function applyAttributes(t, openingElement, componentName, sourceFileName, compo
...
-  if(openingElement.node.attributes.find(node => {
+  if (
+    !ignoredComponentFromOptions
+    && openingElement.node.attributes.find(node => {
     if (!node.name) return
     return node.name.name === elementAttributeName
-  }) == null){
-    const name = openingElement.node.name.name || 'unknown'
-    if (ignoredElements.includes(name)) {
+  }) == undefined){
```
Comment (nfriedly) [inline index.js:146]: You could do `!` instead of comparing to undefined here too. Not a big deal, though.

File: index.js [hunk 8]
```diff
@@ -116,30 +126,44 @@ function applyAttributes(t, openingElement, componentName, sourceFileName, compo
...
-    if (!node.name) return
-    return node.name.name === componentAttributeName
-  }) == null)){
+  if (
+    componentName 
+    && !ignoredComponentFromOptions
+    && (openingElement.node.attributes.find(node => {
+      if (!node.name) return
+      return node.name.name === componentAttributeName
+    }) == undefined)){
```
Comment (nfriedly) [inline index.js:166]: Maybe we could incapsulate this logic into a `hasNodeNamed(openingElement, name)` helper?

**Review by nfriedly — APPROVED** [review]
I left some nitpicky comments below, but honestly, this is perfectly fine to ship as-is. Nice work. 

I especially love all of the tests! ❤️

File: index.js (hunk 8 above)
Comment (sabrina-li) [reply index.js]: Thanks for the recommendations! I updated according to your comments, mind giving this another look through?

**Review by nfriedly — APPROVED** [review]
Nice work!

File: index.js [hunk 9]
```diff
@@ -177,11 +170,11 @@ function applyAttributes(t, openingElement, componentName, sourceFileName, attri
     sourceFileName
     && !ignoredComponentFromOptions
     && (componentName || ignoredElement === false)
-    && openingElement.node.attributes.find(node => {
+    && !openingElement.node.attributes.find(node => {
       if (!node.name) return
       return node.name.name === sourceFileAttributeName
```
Comment (nfriedly) [inline index.js:175]: Oh, you could use the `hasNodeNamed` helper here too, I think.

Comment (nfriedly) [reply index.js:175]: (Sorry, I looked at this earlier but didn't write anything then, and then I forgot a moment ago)

Comment (sabrina-li) [reply index.js]: Ah sorry I missed that one! Thanks for the catch!

**Comment by JoshMiers-FS** [comment]
@nfriedly - can you do a final once-over.

**Review by nfriedly — APPROVED** [review]
Fantastic! 

Ship it! 🚢

**Review by JoshMiers-FS — APPROVED** [review]
LGTM!
